core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,182 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+#define KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
+
+#include <functional>
+
+namespace Kokkos {
+namespace Experimental {
+
+template <typename T>
+class MaybeReferenceCountedPtr {
+ protected:
+  MaybeReferenceCountedPtr(T* value_ptr)
+      : m_value_ptr(value_ptr), m_control(nullptr) {}
+
+  template <class Deleter>
+  MaybeReferenceCountedPtr(T* value_ptr, Deleter deleter)
+      : m_value_ptr(value_ptr), m_control(new Control{std::move(deleter), 1}) {}
+
+ public:
```

Comment:
Unless we are supporting public references/pointers to `MaybeReferenceCountedPtr`, the "Rule of 5" functions can `protected` instead of `public`.

Reply:
I don't think so:
```
/tmp/kokkos_new/core/src/SYCL/Kokkos_SYCL.cpp:80:7: error: field of type 'Kokkos::Experimental::MaybeReferenceCountedPtr<Impl::SYCLInternal>' has protected move constructor
    : m_space_instance(Kokkos::Experimental::UnmanagedPtr<Impl::SYCLInternal>(
      ^
/tmp/kokkos_new/core/src/impl/Kokkos_HostSharedPtr.hpp:75:19: note: declared protected here
  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
                  ^
/tmp/kokkos_new/core/src/SYCL/Kokkos_SYCL.cpp:87:7: error: field of type 'Kokkos::Experimental::MaybeReferenceCountedPtr<Impl::SYCLInternal>' has protected move constructor
    : m_space_instance(Kokkos::Experimental::HostSharedPtr<Impl::SYCLInternal>(
      ^
/tmp/kokkos_new/core/src/impl/Kokkos_HostSharedPtr.hpp:75:19: note: declared protected here
  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
                  ^
2 errors generated.
core/src/CMakeFiles/kokkoscore.dir/build.make:315: recipe for target 'core/src/CMakeFiles/kokkoscore.dir/SYCL/Kokkos_SYCL.cpp.o' failed
```